msal/src/main/java/com/microsoft/identity/client/tenantprofile/AccountAdapter.java

```diff
@@ -0,0 +1,261 @@
+            final List<String> homeAccountIds = new ArrayList<String>() {{
+                for (final ICacheRecord cacheRecord : homeRecords) {
+                    add(cacheRecord.getAccount().getHomeAccountId());
+                }
+            }};
+
+            return homeAccountIds.contains(guestAccountHomeAccountId);
+        }
+
+        @Override
+        public List<ICacheRecord> filter(@NonNull final List<ICacheRecord> records) {
+            final List<ICacheRecord> result = new ArrayList<>();
+
+            // An account is orphaned if:
+            // 1. There is no home account which exists which has the same home_account_id
+
+            // First, get the home accounts....
+            final List<ICacheRecord> homeRecords =
+                    filterCacheRecords(
+                            records,
+                            new HomeAccountFilter(false)
+                    );
+
+            // Then, get the guest accounts...
+            final List<ICacheRecord> guestRecords =
+                    filterCacheRecords(
+                            records,
+                            new HomeAccountFilter(true)
+                    );
+
+            // Iterate over the guest accounts and find those which have no associated home account
+            for (final ICacheRecord guestRecord : guestRecords) {
+                if (hasNoCorrespondingHomeAccount(guestRecord, homeRecords)) {
+                    result.add(guestRecord);
+                }
+            }
+
+            return result;
+        }
+    };
+
+    /**
+     * For a supplied List of ICacheRecords, create each root IAccount based on the home
+     * account and then add child-nodes based on any authorized tenants.
+     *
+     * @param allCacheRecords
+     * @return
+     */
+    @NonNull
+    public static List<IAccount> adapt(@NonNull final List<ICacheRecord> allCacheRecords) {
+        // First, get all of the ICacheRecords for home accounts...
+        final List<ICacheRecord> homeCacheRecords = filterCacheRecords(
+                allCacheRecords,
+                new HomeAccountFilter(false)
+        );
+
+        // Then, get all of the guest accounts...
+        final List<ICacheRecord> guestCacheRecords = filterCacheRecords(
+                allCacheRecords,
+                new HomeAccountFilter(true)
+        );
+
+        final List<ICacheRecord> orphanCacheRecords = filterCacheRecords(
+                allCacheRecords,
+                orphanAccountFilter
+        );
+
+        // Remove the orphan records from the guest records...
+        guestCacheRecords.removeAll(orphanCacheRecords);
+
+        final List<IAccount> rootAccounts = createRootAccounts(homeCacheRecords);
+        appendChildren(rootAccounts, guestCacheRecords);
+        rootAccounts.addAll(createOrphanRoots(orphanCacheRecords));
+
+        return rootAccounts;
+    }
+
+    private static List<IAccount> createOrphanRoots(
```

Comment:
These likely are not "roots" since they are orphaned.... I think we can revisit this naming concern in the future though.

Reply:
The orphans do have a root - a sort of 'stub' `MultiTenantAccount` that has an id, but no claims. We can work on the naming here

Reply:
Renamed this method to `createIAccountsForGuestsNotSignedIntoHomeTenant` in af766f7ba4aaf018ef97a2fe592dc50703a65342